engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/NodeCompiler.scala

```diff
@@ -32,9 +31,47 @@ import shapeless.Typeable
 import shapeless.syntax.typeable._
 import cats.instances.list._
 import cats.implicits.toTraverseOps
+import pl.touk.nussknacker.engine.compile.NodeTypingInfo.DefaultExpressionId
+import pl.touk.nussknacker.engine.compile.nodecompilation.NodeCompiler.{ExpressionTypingResult, FieldExpressionTypingResult, NodeCompilationResult}
 
 import scala.util.{Failure, Success, Try}
 
+object NodeCompiler {
+
+  case class NodeCompilationResult[T](expressionTypingInfo: Map[String, ExpressionTypingInfo],
+                                      parameters: Option[List[Parameter]],
+                                      validationContext: ValidatedNel[ProcessCompilationError, ValidationContext],
+                                      compiledObject: ValidatedNel[ProcessCompilationError, T],
+                                      expressionType: Option[TypingResult] = None) {
+    def errors: List[ProcessCompilationError] = (validationContext.swap.toList ++ compiledObject.swap.toList).flatMap(_.toList)
+
+    def withVariable(varName: String)(implicit nodeId: NodeId): NodeCompilationResult[T] = copy(
+      validationContext = validationContext.andThen(_.withVariable(varName, expressionType.getOrElse(Unknown)))
```

Comment:
I wouldn't put this method here, but do validation in appropriate parts of NodeCompiler (maybe we shall add more specific methods like compileVariableBuilder?). Otherwise it's NodeCompiler caller responsibility to invoke it, and we can forget about it - e.g. in NodeDataValidator it's not invoked, and validation is not full